File: app/src/main/java/com/tcc/sicv/base/ResultListenerFactory.java
package com.tcc.sicv.base;

import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;

import com.tcc.sicv.data.model.FlowState;

import static com.tcc.sicv.data.model.Status.ERROR;
import static com.tcc.sicv.data.model.Status.SUCCESS;

public class ResultListenerFactory<T> {
   public Result<T> create(@NonNull final MutableLiveData<FlowState<T>> liveData){
       return new Result<T>() {
           @Override
           public void onSuccess(T data) {
               liveData.postValue(new FlowState<>(data, null, SUCCESS));
           }

           @Override
           public void onFailure(Throwable throwable) {
                liveData.postValue(new FlowState<T>(null, throwable, ERROR));
           }
       };
   }
}

File: app/src/main/java/com/tcc/sicv/base/ResultListenerFactory.kt
package com.tcc.sicv.base

import android.arch.lifecycle.MutableLiveData

import com.tcc.sicv.data.model.FlowState

import com.tcc.sicv.data.model.Status.ERROR
import com.tcc.sicv.data.model.Status.SUCCESS

class ResultListenerFactory<T> {
    fun create(liveData: MutableLiveData<FlowState<T>>): Result<T> {
        return object : Result<T> {
            override fun onSuccess(data: T) {
                liveData.postValue(FlowState(data, null, SUCCESS))
            }

            override fun onFailure(throwable: Throwable?) {
                liveData.postValue(FlowState(null, throwable, ERROR))
            }
        }
    }
}
